### PR TITLE
Nonzero exit code when diff is not clean

### PR DESCRIPTION
**Pull request by yannickhilber**
Store diff prints to text buffer and exit program with code 2 when the diff is not empty.

### CODE REVIEW

**Review by ruuda — COMMENTED** [review]
This works, but it is also quite invasive. What I would do instead is to keep track of something like `has_changes = False` in `main`, then make `print_diff` return whether there was a change, then we can do something like

```python
has_changes |= print_diff(...)
```
and then at the end we can check `has_changes`.

(Be careful with `or` and side effects as `or` short-circuits; if the left-hand side is true then it never executes the right-hand side. `|=` does not short-circuit.)

**Comment by yannickhilber** [comment]
You're right this solution is cleaner. It required to track the changes on `print_team_members_diff` where we instantiate a `Diff` and when the we check if the organizations matches.

I needed to add a boolean variable in `print_diff` and `print_team_members_diff` to go through all the if statements. I called it `has_diff` to match with the function names and not to confuse it with `has_changes` in the main function.

File: main.py [hunk 1]
```diff
@@ -877,7 +877,11 @@ def print_diff(
         header_to_remove: str,
         header_to_change: str,
     ) -> None:
```
Comment (ruuda) [inline main.py:879]:
```suggestion
    ) -> bool:
```
We should update the type signature to indicate that the function returns a boolean now. You can typecheck the file with `mypy --strict`.

File: main.py [hunk 2]
```diff
@@ -912,28 +921,39 @@ def print_team_members_diff(
     target_members: Set[TeamMember],
     actual_members: Set[TeamMember],
 ) -> None:
```
Comment (ruuda) [inline main.py:923]:
```suggestion
) -> bool:
```
Also here.

**Review by ruuda — COMMENTED** [review]
This looks good! Only please make sure the file typechecks, you can typecheck it with `mypy --strict`.

**Comment by yannickhilber** [comment]
> This looks good! Only please make sure the file typechecks, you can typecheck it with `mypy --strict`.

Thanks I was not aware of this tool. I will use it in the future.

**Comment by yannickhilber** [comment]
> Hmm something weird happened, this pull request now touches `LICENSE` and `nixpkgs-pinned.nix`, and I see some commits that I authored but you committed ... I think you rebased too much. You could fix this with
> 
> ```
> git fetch origin
> git rebase --onto origin/master 3_nonzero_exit_code_diff^ 3_nonzero_exit_code_diff
> ```

This is not the first time that I rebased too much and past commits are never added. I tried to reproduce what happened by adding a commit and squashing it to the previous with `git rebase -i HEAD~8` but it didn't work :confused: 

Thanks a lot for the fix, it worked perfectly !

**Pull request merged**